### PR TITLE
Refactor place scenery

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Its been annoying me for quite a while this function with how nested it was. I've removed as much of the nesting as possible and named variables where appropriate.

### CODE REVIEW

**Comment by IntelOrca** [comment]
It would be nice to ditch all the register names as well.

File: src/world/map.c [hunk 1]
```diff
@@ -3014,191 +3014,240 @@ void game_command_place_scenery(int* eax, int* ebx, int* ecx, int* edx, int* esi
...
+		*ebx = MONEY32_UNDEFINED;
+		return;
+	}
+
+	if (!map_check_free_elements_and_reorganise(1)) {
+		*ebx = MONEY32_UNDEFINED;
+		return;
+	}
+
+	if (!(byte_9D8150 & 1) && (x > gMapSizeMaxXY || y > gMapSizeMaxXY)) {
```
Comment (marijnvdwerf) [inline src/world/map.c]: `byte_9D8150` probably should be a `bool`.

image

File: src/world/map.c [hunk 2]
```diff
@@ -3014,191 +3014,240 @@ void game_command_place_scenery(int* eax, int* ebx, int* ecx, int* edx, int* esi
...
+		*ebx = MONEY32_UNDEFINED;
+		return;
+	}
+
+	rct_scenery_entry* scenery_entry = (rct_scenery_entry*)object_entry_groups[OBJECT_TYPE_SMALL_SCENERY].chunks[scenery_type];
```
Comment (marijnvdwerf) [inline src/world/map.c]: `get_small_scenery_entry`?

File: src/world/map.c [hunk 3]
```diff
@@ -3014,191 +3014,240 @@ void game_command_place_scenery(int* eax, int* ebx, int* ecx, int* edx, int* esi
...
+		footpath_remove_litter(x, y, targetHeight);
+		if(!gCheatsDisableClearanceChecks && (scenery_entry->small_scenery.flags & SMALL_SCENERY_FLAG_ALLOW_WALLS)) {
+			map_remove_walls_at(x, y, targetHeight, targetHeight + scenery_entry->small_scenery.height);
+		}
+	}
+
+	rct_map_element* surface_element = map_get_surface_element_at(x / 32, y / 32);
+
+	if(!gCheatsDisableClearanceChecks && (surface_element->properties.surface.terrain & 0x1F)){
+		int water_height = ((surface_element->properties.surface.terrain & 0x1F) * 16) - 1;
```
Comment (marijnvdwerf) [inline src/world/map.c]: `MAP_ELEMENT_WATER_HEIGHT_MASK`

File: src/world/map.c [hunk 4]
```diff
@@ -3014,191 +3014,240 @@ void game_command_place_scenery(int* eax, int* ebx, int* ecx, int* edx, int* esi
...
+		*ebx = MONEY32_UNDEFINED;
+		return;
+	}
+
+	if (!gCheatsDisableSupportLimits && 
+		!(scenery_entry->small_scenery.flags & SMALL_SCENERY_FLAG18) && 
+		supportsRequired) {
+
+		if(!isOnWater){
+			if((surface_element->properties.surface.terrain & 0x1F) ||
```
Comment (marijnvdwerf) [inline src/world/map.c]: `MAP_ELEMENT_WATER_HEIGHT_MASK`

File: src/world/map.c [hunk 5]
```diff
@@ -3014,191 +3014,240 @@ void game_command_place_scenery(int* eax, int* ebx, int* ecx, int* edx, int* esi
...
+		}else{
+			gGameCommandErrorText = STR_CAN_ONLY_BUILD_THIS_ON_LAND;
+			*ebx = MONEY32_UNDEFINED;
+			return;
+		}
+	}
+
+	int bp = quadrant;
+	int zLow = targetHeight / 8;
+	int zHigh = zLow + ((scenery_entry->small_scenery.height + 7) / 8);
```
Comment (marijnvdwerf) [inline src/world/map.c]: `ceil2(scenery_entry->small_scenery.height, 8) / 8` might make this a bit more obvious?

File: src/world/map.c [hunk 6]
```diff
@@ -3014,191 +3014,240 @@ void game_command_place_scenery(int* eax, int* ebx, int* ecx, int* edx, int* esi
...
+
+	if(!(gScreenFlags & SCREEN_FLAGS_SCENARIO_EDITOR) && 
+		!gCheatsSandboxMode && 
+		!map_is_location_owned(x, y, targetHeight)){
+
+		*ebx = MONEY32_UNDEFINED;
+		return;
+	}
+
+	if(flags & GAME_COMMAND_FLAG_APPLY && !(flags & 0x40)){
```
Comment (marijnvdwerf) [inline src/world/map.c]: `GAME_COMMAND_FLAG_GHOST`?

File: src/world/map.c [hunk 7]
```diff
@@ -3014,191 +3014,240 @@ void game_command_place_scenery(int* eax, int* ebx, int* ecx, int* edx, int* esi
...
+	type |= rotation;
+	new_map_element->type = type;
+	new_map_element->properties.scenery.type = scenery_type;
+	new_map_element->properties.scenery.age = 0;
+	new_map_element->properties.scenery.colour_1 = colour1;
+	new_map_element->properties.scenery.colour_2 = colour2;
+	new_map_element->clearance_height = new_map_element->base_height + ((scenery_entry->small_scenery.height + 7) / 8);
+	
+	if(supportsRequired){
+		new_map_element->properties.scenery.colour_1 |= 0x20;
```
Comment (marijnvdwerf) [inline src/world/map.c]: Maybe add a constant for this flag? Is also used by the scenery painting

File: src/world/map.c (hunk 1 above)
Comment (duncanspumpkin) [inline src/world/map.c]: I can't work out why this is used. As far as I can tell it's an override for when the track window but not sure why you would want to place outside of the x y limits during this but not on any other scenery

File: src/world/map.c (hunk 5 above)
Comment (IntelOrca) [inline src/world/map.c]: Why do you still have the `/8`?

Comment (marijnvdwerf) [inline src/world/map.c]: Because zLow and zHeight are both in units?

**Comment by marijnvdwerf** [comment]
A normal ceil would be AND'ing with ~0b111, right?
Op wo 16 nov. 2016 om 19:04 schreef Ted John <notifications@github.com>

**Pull request merged**